File: code_snippets/go/usecases/eventprocessing/eventenrichment/packagetracker.go
package main

import (
	"context"
	"errors"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
)

type LocationUpdate struct {
	Timestamp string `json:"timestamp"`
	Location  string `json:"location"`
}

type PackageInfo struct {
	FinalDestination string           `json:"finalDestination"`
	Locations        []LocationUpdate `json:"locations"`
}

// <start_here>
// <mark_2>
type PackageTracker struct{}

// </mark_2>

// <mark_2>
// <mark_3>
func (PackageTracker) RegisterPackage(ctx restate.ObjectContext, packageInfo PackageInfo) error {
	// </mark_2>
	// </mark_3>
	// <mark_1>
	restate.Set[PackageInfo](ctx, "package-info", packageInfo)
	// </mark_1>
	return nil
}

// <mark_3>
// <mark_2>
func (PackageTracker) UpdateLocation(ctx restate.ObjectContext, locationUpdate LocationUpdate) error {
	// </mark_2>
	// </mark_3>
	// <mark_1>
	packageInfo, err := restate.Get[*PackageInfo](ctx, "package-info")
	// </mark_1>
	if err != nil {
		return err
	}
	if packageInfo == nil {
		return restate.TerminalError(errors.New("package not found"))
	}

	if packageInfo.FinalDestination == "" {
		return restate.TerminalError(errors.New("package not found"))
	}

	packageInfo.Locations = append(packageInfo.Locations, locationUpdate)
	// <mark_1>
	restate.Set[PackageInfo](ctx, "package-info", *packageInfo)
	// </mark_1>
	return nil
}

// <mark_3>
// <mark_2>
func (PackageTracker) GetPackageInfo(ctx restate.ObjectSharedContext) (*PackageInfo, error) {
	// </mark_2>
	// </mark_3>
	// <mark_1>
	return restate.Get[*PackageInfo](ctx, "package-info")
}

func main() {
	if err := server.NewRestate().
		Bind(restate.Reflect(PackageTracker{})).
		Start(context.Background(), ":9080"); err != nil {
		log.Fatal(err)
	}
}

// <end_here>

File: code_snippets/java/src/main/java/usecases/eventprocessing/eventenrichment/PackageTracker.java
package usecases.eventprocessing.eventenrichment;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.SharedObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import dev.restate.sdk.types.StateKey;
import dev.restate.sdk.types.TerminalException;
import usecases.eventprocessing.eventenrichment.types.LocationUpdate;
import usecases.eventprocessing.eventenrichment.types.PackageInfo;

// <start_here>
// <mark_2>
@VirtualObject
public class PackageTracker {
  // </mark_2>

  private static final StateKey<PackageInfo> PACKAGE_INFO =
      StateKey.of("package-info", PackageInfo.class);

  // <mark_3>
  // <mark_2>
  @Handler
  public void registerPackage(ObjectContext ctx, PackageInfo packageInfo) {
    // </mark_2>
    // </mark_3>
    // <mark_1>
    ctx.set(PACKAGE_INFO, packageInfo);
    // </mark_1>
  }

  // <mark_3>
  // <mark_2>
  @Handler
  public void updateLocation(ObjectContext ctx, LocationUpdate locationUpdate) {
    // </mark_2>
    // </mark_3>
    // <mark_1>
    var packageInfo =
        ctx.get(PACKAGE_INFO)
            // </mark_1>
            .orElseThrow(() -> new TerminalException("Package not found"));

    packageInfo.addLocation(locationUpdate);
    // <mark_1>
    ctx.set(PACKAGE_INFO, packageInfo);
    // </mark_1>
  }

  // <mark_3>
  // <mark_2>
  @Shared
  public PackageInfo getPackageInfo(SharedObjectContext ctx) {
    // </mark_2>
    // </mark_3>
    // <mark_1>
    return ctx.get(PACKAGE_INFO)
        // </mark_1>
        .orElseThrow(() -> new TerminalException("Package not found"));
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new PackageTracker()));
  }
}
// <end_here>
